subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/MerkleDirectorySnapshotBuilder.java

```diff
@@ -46,42 +43,19 @@ public static MerkleDirectorySnapshotBuilder noSortingRequired() {
 
     private MerkleDirectorySnapshotBuilder(boolean sortingRequired) {
         this.sortingRequired = sortingRequired;
+        addLevel();
     }
 
-    public void preVisitDirectory(String absolutePath, String name) {
-        relativePathSegmentsTracker.enter(name);
-        levelHolder.addLast(new ArrayList<>());
-        directoryAbsolutePaths.addLast(absolutePath);
-    }
-
-    @Override
-    public boolean preVisitDirectory(CompleteDirectorySnapshot directorySnapshot) {
-        preVisitDirectory(directorySnapshot.getAbsolutePath(), directorySnapshot.getName());
-        return true;
-    }
-
-    @Override
-    public void visitFile(CompleteFileSystemLocationSnapshot fileSnapshot) {
-        if (relativePathSegmentsTracker.isRoot()) {
-            result = fileSnapshot;
-        } else {
-            levelHolder.peekLast().add(fileSnapshot);
-        }
-    }
-
-    @Override
-    public void postVisitDirectory(CompleteDirectorySnapshot directorySnapshot) {
-        postVisitDirectory(true, directorySnapshot.getAccessType());
+    public void enterDirectory() {
+        addLevel();
     }
 
-    public void postVisitDirectory(AccessType accessType) {
-        postVisitDirectory(true, accessType);
+    public void visitLeafElement(FileSystemLeafSnapshot snapshot) {
+        collectResult(snapshot);
     }
 
-    public boolean postVisitDirectory(boolean includeEmpty, AccessType accessType) {
-        String name = relativePathSegmentsTracker.leave();
+    public boolean leaveDirectory(boolean includeEmpty, AccessType accessType, String absolutePath, String name) {
```

Comment:
Do you think it is already worth creating an enum for `includeEmpty`?

Reply:
Added.